### DISCUSSION TITLE
Storage backend migration script

### DISCUSSION THREAD

**Discussion by felixjogris** [PrivateBin/PrivateBin#1007] [Ideas]
Hi,

I started a storage backend migration script (https://github.com/felixjogris/PrivateBin/blob/migrate/migrate.php). It copies pastes, attachments, and comments from one backend (e.g. file system) to another one (e.g. a database). Both backends may be of same type, e.g. from database to database.
Before submitting a pull request, I'd like to discuss a few points:

- Is this command line script welcome?
- Better add a member function for listing all paste ids to AbstractData and derived classes, or better stick with reflection to gain access to private properties like migrate.php currently does?
- Better change all static properties (e.g. $_db or $_path) in data backend classes to non-static ones, or let migrate.php first read all pastes and comments to memory, and then write to the new storage backend?

Best,
Felix

**Comment by elrido** [comment]
Very nice work and a creative use of reflection! It parses flags and has a help message. It is certainly an enhancement that has been requested, see #359, so it would make sense to add it.

I would recommend to:
- reg. point 2: I'd recommend adding a new (public) member function - the `_getExpiredPastes` is already close to that, but has extra steps filtering the result and limits itself to a batch size, so it would be a simplified version of that
- reg. point 3: those static properties would be shared when we'd have two backends of the same type in play - since PrivateBin in normal operation will try and share the backend object via [DI](https://en.wikipedia.org/wiki/Dependency_injection), making these properties non-static shouldn't impact memory usage, so I'd recommend changing it to non-static.
- assuming this script is used with larger instances that have hundreds or thousands of pastes and comments, it would probably be a bit slower, but more memory conserving, if the pastes and comments get converted one at a time (folding lines 53-58 into the loop starting at 44) instead of compiling an in-memory representation of all pastes and comments before storing them all
- it would be nice to have a unit test for this, at least a simple one that converts a paste with comment from file storage to sqlite in memory and back, validating the data is complete at each stage - I'm willing to write one
- create a new directory called `bin` and place it there
- update the [INSTALL.md](https://github.com/PrivateBin/PrivateBin/blob/master/INSTALL.md#changing-the-path) to mention that `bin` should also live outside the document root
- update the CREDITS.md and CHANGELOG.md, attributing your effort
- (on our side) also move the [two](https://github.com/PrivateBin/PrivateBin/blob/master/tst/ConfigurationTestGenerator.php) [other](https://github.com/PrivateBin/PrivateBin/blob/master/tst/IconTest) scripts into `bin`

**Reply by elrido** [reply]
Having merged your script in #1012, @felixjogris, I've now raised #1014 with the move of the other two scripts and the unit test. It does indeed work nicely and just as advertised! :-D Could you have a look at the PR, please?

I did discover that the third argument in `getopt` requires PHP >= 7.1, so I added a version check to the script. Also made it executable (it still works when called with `php` explicitly).
